### ISSUE TITLE
An empty directory remains after the uninstall

### ISSUE THREAD

**Issue by geghamyan** [oleg-shilo/wixsharp#464]
Steps to reproduce:
* Use a real certificate with the "Certificate" sample 
* Install/uninstall the MSI
* Observe that an empty "%ProgramFiles%\My Company" remains after the uninstall

The issue is related to components without files or registry elements, in this particular case a component with the Certificate element. The WixSharp already contains a workaround code to inject "RemoveFolder" for a parent directory of the component, but it looks like it's not enough, it should be done for all chain of parent folders.

**Comment by oleg-shilo** [comment]
Hi Tigran,

There may be some dev collision between us as I have done some changes in the same area as you did for your PR. The changes are in master but not released yet. 

Simply when I tested the certificate sample against the current codebase the all user folders were deleted. Thus it is either because of the changes I made or because of the difference in our environments. 

Can you please test the attached sample in your environment? I have borrowed `oracle.cer` from VirtualBox for it.
[Certificates.zip](https://github.com/oleg-shilo/wixsharp/files/2342189/Certificates.zip)

**Comment by geghamyan** [comment]
Hi Oleg,

The attached sample works, but it works is because you've removed the certificates :)
The component without files or registries (the one which contains "Certificate") causes this behavior.
```
            <Component Id="NameOfCert" Guid="6f330b47-2577-43ad-9095-1861c5cc97d9" KeyPath="yes">
              <Certificate Id="NameOfCert" BinaryKey="mycert" Name="NameOfCert" StoreLocation="localMachine" StoreName="personal" xmlns="http://schemas.microsoft.com/wix/IIsExtension" />
              <RemoveFolder Id="INSTALLDIR" On="uninstall" />
            </Component>
```
You've already put quite a lot of code to do workaround this issue, i.e. "RemoveFolder" is inserted in this component by the following code in `AutoElement.cs`
```
                    var componentsWithNoFilesOrRegestry = dirComponents.Where(x => !x.ContainsFilesOrRegistries()).ToArray();

                    foreach (XElement item in componentsWithNoFilesOrRegestry)
                    {
                        //if (!item.Attribute("Id").Value.EndsWith(".EmptyDirectory"))
                        EnsureKeyPath(item);

                        if (!xDir.ContainsAnyRemoveFolder())
                            InsertRemoveFolder(xDir, item, "uninstall"); //to keep WiX/compiler happy and allow removal of the dummy directory
                    }
```
Thus the inserted `RemoveFolder` takes care for INSTALLDIR folder, but the **parent** `ProgramFilesFolder.My_Company` folder doesn't get deleted after the installation.
It looks like the a component without files or registry entries prevents all parent folders of the component from deleting.

What I did was: 
- In the function `CreateEmptyComponentsInDirectoriesToRemove()` identified all directories with components with no files or registry. 
- Create new empty components in all parent directories 
- Later the piece of code which inserts `RemoveFolder` finds all the newly inserted empty components and inserts `RemoveFolder` for the component's directory.
- The PR also contains a fix for the empty folders handling. The `GenerateComponentId` functions inserts the suffix before the id, but the code in the repo expects the suffix at the end `EndsWith(".EmptyDirectory")`. I've rearranged suffix to be at the end of the component id.
- Things that I noticed but didn't fix: not all the code uses `GenerateComponentId` to generate unique components ids, for example function `CrteateComponentFor`  or code in `WixSharp.Compiler.ProcessDirectory` which takes care of empty directories.

Here is the Certificates sample with the OpenVPN cert.
[Certificates.zip](https://github.com/oleg-shilo/wixsharp/files/2342746/Certificates.zip)
